pyomo/contrib/mindtpy/algorithm_base_class.py

```diff
@@ -1626,73 +1527,67 @@ def set_up_lazy_OA_callback(self, mainopt):
             The MIP solver.
         """
         if self.config.mip_solver == 'cplex_persistent':
-            lazyoa = mainopt._solver_model.register_callback(
+            lazyoa = self.mip_opt._solver_model.register_callback(
                 single_tree.LazyOACallback_cplex
             )
             # pass necessary data and parameters to lazyoa
             lazyoa.main_mip = self.mip
-            self.solve_data = MindtPySolveData()
-            self.export_attributes()
-            lazyoa.solve_data = self.solve_data
             lazyoa.config = self.config
-            lazyoa.opt = mainopt
-            mainopt._solver_model.set_warning_stream(None)
-            mainopt._solver_model.set_log_stream(None)
-            mainopt._solver_model.set_error_stream(None)
+            lazyoa.opt = self.mip_opt
+            lazyoa.mindtpy_object = self
+            self.mip_opt._solver_model.set_warning_stream(None)
+            self.mip_opt._solver_model.set_log_stream(None)
+            self.mip_opt._solver_model.set_error_stream(None)
         if self.config.mip_solver == 'gurobi_persistent':
-            mainopt.set_callback(single_tree.LazyOACallback_gurobi)
+            self.mip_opt.set_callback(single_tree.LazyOACallback_gurobi)
+            self.mip_opt.mindtpy_object = self
+            self.mip_opt.config = self.config
 
     ##########################################################################################################################################
     # mip_solve.py
 
-    def solve_main(self, config):
+    def solve_main(self):
         """This function solves the MIP main problem.
 
-        Parameters
-        ----------
-        config : ConfigBlock
-            The specific configurations for MindtPy.
-
         Returns
         -------
         self.mip : Pyomo model
             The MIP stored in self.
         main_mip_results : SolverResults
             Results from solving the main MIP.
         """
+        config = self.config
         self.mip_iter += 1
 
         # setup main problem
-        self.setup_main(config)
-        mainopt, mip_args = self.set_up_mip_solver(config)
+        self.setup_main()
+        mip_args = self.set_up_mip_solver()
```

Comment:
Good idea. Actually, we also need to change the name of `set_up_mip_solver` since we only generate the arguments here.

Reply:
Sorry to be terse: By "there's no point," I meant that you only get performance benefits from the persistent solver when you call `set_instance` once, and then (in the case of everything except appsi) tell the solver what changes you made to the model every iteration (for example, adding cuts). So I'm not saying there's something *wrong* with this code. But I think it may be confusing to users that if they specify a persistent solver as the MIP solver, they aren't actually getting the benefits of faster solves because the way you're doing this, the solver doesn't know what changed on the model: Instead, it starts the solve from scratch every time whether it's a persistent solver or not.

Reply:
Yes. We haven't made effective use of `persistent_solver` to gain performance benefits here. The thing is we need to use the `add_constraint` and `add_var` instead of directly defining them when using `persistent solver`. Compared to `appsi_solver`, incrementally updating the model takes more effort. Any suggestions for the implementation?

Reply:
Personally, I think it would be fine to not support the original persistent solvers and just support appsi. I would prefer that to allowing users to specify persistent solvers, but then not really using them as such. But it's up to what you and @bernalde think.